README.md
audl
=========
[![Build Status](https://travis-ci.org/tadachi/audl.svg?branch=master)](https://travis-ci.org/tadachi/audl)

#### Description

Quickly and conveniently download music, podcasts or audio from Youtube.

Has options for batch and checking bitrate of Youtube audio.

Special thanks to [fent](https://github.com/fent/node-ytdl). His work is a core part of audl.

#### Install

```bash
npm -g install audl
```

#### Usage

##### Single download
```
audl -d https://www.youtube.com/watch?v=9bZkp7q19f0
downloading [                    ] 100% 0.0s 4007494/4007494 - psy_gangnam_style_m_v.m4a
```
##### Batch download
```
audl -b batch.txt

downloading [                    ] 100% 0.0s 12386745/12386745 - the_epic_final_fantasy_vi_medley.m4a
downloading [                    ] 100% 0.0s 17081058/17081058 - final_fantasy_epic_orchestral_medley.m4a
downloading [                    ] 100% 0.0s 16404793/16404793 - the_epic_final_fantasy_v_medley.m4a
```
##### Single getinfo
```
audl -i https://www.youtube.com/watch?v=N1QgLqWb0rE

title                             itag  encoding  bitrate
--------------------------------  ----  --------  -------
the_epic_final_fantasy_vi_medley  140   aac       128
```
##### Batch getinfo
```
audl -I batch.txt

title                                 itag  encoding  bitrate  url
------------------------------------  ----  --------  -------  -------------------------------------------
the_epic_final_fantasy_vi_medley      140   aac       128      https://www.youtube.com/watch?v=N1QgLqWb0rE
final_fantasy_epic_orchestral_medley  140   aac       128      https://www.youtube.com/watch?v=1smhVzu_Y7I
the_epic_final_fantasy_v_medley       140   aac       128      https://www.youtube.com/watch?v=Ginfs4TY0oo
```
##### Help
```
  audl -h

  Usage: index [options]

  audl - A convenient node command-line app to download youtube audio content such as podcasts and music.

            Examples:

            audl -d https://www.youtube.com/watch?v=9bZkp7q19f0
            audl -i https://www.youtube.com/watch?v=9bZkp7q19f0
            audl -b batch.txt
            audl -I batch.txt

  Options:

    -h, --help               output usage information
    -V, --version            output the version number
    -d, --url [url]          Specify youtube link to download.
    -b, --batch [file]       Specify a batch text file of youtube urls (LR separated) and download them al
    -i, --info [url]         Get list of quality options for that youtube content
    -I, --batch_info [file]  Specify a batch text file and get audio quality info of all youtube urls
```

###### Batch Files

See [this](https://github.com/tadachi/audl/blob/master/tests/batch.txt) for an example of a proper batch file for audl.

Each URL is line-feed separated.

##### Tests

To manually test, run these commands.

```
  npm run test_download
  npm run test_getinfo
  npm run test_batch
  npm run test_batch_getinfo
```

##### License

MIT License
 


CHANGELOG.md
##### 0.2.4
- Fixed bug with using hardcoded 'batch.txt', you can now specify any text file in any location.

##### 0.2.3
- Change single download, single getinfo to use promises.

##### 0.2.2
- Cleaned up -h help text
- Updated README.md with run examples
- Removed -q quality option which might be implemented in the future depending on youtube improving bitrate audio quality.

##### 0.2.1
- Minor bugfixes

##### 0.2.0
- Major bugfixes to all options -d -i -b -I
- Implemented promises for batch options

##### 0.1.0
Initial version

index.js
#!/usr/bin/env node
"use strict";
exports.__esModule = true;
var ytdl = require("ytdl-core");
var program = require("commander");
var fs = require("fs");
var ProgressBar = require("progress");
var Promise = require("bluebird");
require('console.table');
/*
 * Takumi Adachi
 * 02/01/2017
 *
 * MIT License
 *
 * Node application to quickly and conveniently download adequate quality music, podcasts or audio from Youtube.
 *
 * Has options for batch and checking bitrate of youtube audio.
 */
// Run main routine,
main();
// Main routine.
function main() {
    program
        .version('0.2.4')
        .description("audl - A convenient node command-line app to download youtube audio content such as podcasts and music.\n\n            Examples:\n\n            audl -d https://www.youtube.com/watch?v=9bZkp7q19f0\n            audl -i https://www.youtube.com/watch?v=9bZkp7q19f0\n            audl -b batch.txt\n            audl -I batch.txt")
        .option('-d, --url [url]', 'Specify youtube link to download.')
        .option('-b, --batch [file]', 'Specify a batch text file of youtube urls (LR separated) and download them all')
        .option('-i, --info [url]', 'Get list of quality options for that youtube content')
        .option('-I, --batch_info [file]', 'Specify a batch text file and get audio quality info of all youtube urls')
        .parse(process.argv);
    // Check if url is a valid youtube link.
    function valid_youtube_match(url) {
        var youtube_valid_regexp = /^.*(youtu.be\/|v\/|u\/\w\/|embed\/|watch\?v=|\&v=|\?v=)([^#\&\?]*).*/;
        var match = url.match(youtube_valid_regexp);
        if (match && match[2].length == 11) {
            return true;
        }
        else {
            return false;
        }
    }
    // Download youtube content as audio.
    if (program.url) {
        var url = program.url;
        if (!valid_youtube_match(url)) {
            console.log(program.url + 'is not a valid youtube url.');
            return;
        }
        console.log();
        YTdownloadAsAudio(url).then(function (result) {
        }).error(function (err) {
            console.log(err);
        });
        return;
    }
    // Get itag info and output to console.
    if (program.info) {
        var url = program.info;
        // Check if valid url.
        if (!valid_youtube_match(url)) {
            console.log(url + 'is not a valid youtube url.');
            return;
        }
        getInfo(url).then(function (info) {
            var audio_file_meta = new YTAudioFileMeta(info);
            var title = info.title.replace(/[^a-z]+/gi, '_').toLowerCase(); // music_title.
            var formats = audio_file_meta.formats;
            var itag_info = [];
            function push(itag) {
                itag_info.push({
                    title: title,
                    itag: formats[itag].itag.toString(),
                    encoding: formats[itag].audioEncoding.toString(),
                    bitrate: formats[itag].audioBitrate.toString()
                });
            }
            for (var itag in formats) {
                var found = false;
                if (itag === '139')
                    push(itag);
                if (itag === '140')
                    push(itag);
                if (itag === '141')
                    push(itag);
            }
            console.log();
            console.table(itag_info);
        }).error(function (err) {
            console.log(err);
        });
        return;
    }
    // Get Itag info for a batch of youtube urls.
    if (program.batch_info) {
        // Example: Read from a file of youtube links separated by linefeeds.
        var youtube_urls = fs.readFileSync(program.batch_info).toString().split("\r");
        var getInfo_promises = [];
        var itag_info_1 = [];
        var _loop_1 = function (url) {
            // Remove \r \n from string before validating.
            url = url.replace(/(\r\n|\n|\r)/gm, "");
            if (!valid_youtube_match(url)) {
                console.log(url + ' is not a valid youtube url.');
                return "continue";
            }
            var promise = getInfo(url).then(function (info) {
                var audio_file_meta = new YTAudioFileMeta(info);
                var title = info.title.replace(/[^a-z]+/gi, '_').toLowerCase(); // music_title.
                var formats = audio_file_meta.formats;
                function push(itag) {
                    itag_info_1.push({
                        title: title,
                        itag: formats[itag].itag.toString(),
                        encoding: formats[itag].audioEncoding.toString(),
                        bitrate: formats[itag].audioBitrate.toString(),
                        url: url
                    });
                }
                for (var itag in formats) {
                    if (itag === '139')
                        push(itag);
                    if (itag === '140')
                        push(itag);
                    if (itag === '141')
                        push(itag);
                }
                return itag_info_1;
            });
            getInfo_promises.push(promise);
        };
        for (var _i = 0, youtube_urls_1 = youtube_urls; _i < youtube_urls_1.length; _i++) {
            var url = youtube_urls_1[_i];
            _loop_1(url);
        }
        Promise.map(getInfo_promises, function (info) {
            // Do only one request at a time using concurrency.
        }, { concurrency: 1 }).then(function () {
            console.log();
            console.table(itag_info_1);
        }).error(function (err) {
            console.log(err);
        });
        ;
        return;
    }
    // Download a batch of youtube urls in audio format.
    if (program.batch) {
        // Example: Read from a file of youtube links separated by linefeeds.
        var youtube_urls = fs.readFileSync(program.batch).toString().split("\n");
        var download_promises = [];
        var itag_info = [];
        for (var _a = 0, youtube_urls_2 = youtube_urls; _a < youtube_urls_2.length; _a++) {
            var url = youtube_urls_2[_a];
            // Remove \r \n from string before validating.
            url = url.replace(/(\r\n|\n|\r)/gm, "");
            // Check if valid youtube url.
            if (!valid_youtube_match(url)) {
                console.log(url + 'is not a valid youtube url and will not be downloaded.');
                continue;
            }
            var promise = YTdownloadAsAudio(url);
            download_promises.push(promise);
        }
        Promise.map(download_promises, function (result) {
            // Do only one request at a time using concurrency.
        }, { concurrency: 1 }).then(function (result) {
        }).error(function (err) {
            console.log(err);
        });
        return;
    }
    // If program was called with no arguments, show help.
    if (program.args.length === 0)
        program.help();
    // Return a promise to get youtube info such as itag, quality, bitrate.
    function getInfo(url) {
        return new Promise(function (resolve, reject) {
            ytdl.getInfo(url, function (err, info) {
                if (info)
                    resolve(info);
                if (err)
                    reject(err);
            });
        });
    }
    // Return a promise to download youtube audio content.
    function YTdownloadAsAudio(url) {
        return new Promise(function (resolve, reject) {
            ytdl.getInfo(url, function (err, info) {
                var audio_file_meta = new YTAudioFileMeta(info);
                var file_type = '.m4a';
                // Remove unneeded characters and replace with underscores for readability and make it file friendly.
                var file_name = audio_file_meta.title.replace(/[^a-z]+/gi, '_').toLowerCase() + file_type; // music_title.m4a
                var write_stream = fs.createWriteStream(file_name);
                var audio = ytdl(url, { quality: 140 });
                audio.pipe(write_stream);
                audio.on('response', function (res) {
                    var dataRead = 0;
                    var totalSize = parseInt(res.headers['content-length']);
                    var options = {
                        complete: '\u001b[42m \u001b[0m',
                        incomplete: '\u001b[41m \u001b[0m',
                        width: 20,
                        total: totalSize
                    };
                    //Example download youtube audio and save as .m4a with a progressbar.
                    var bar = new ProgressBar(" downloading [:bar] :percent :etas :current/:total - " + file_name, options);
                    res.on('data', function (data) {
                        var chunk = data.length;
                        bar.tick(chunk);
                    });
                });
                audio.on('finish', function () {
                    resolve(true);
                });
                audio.on('error', function () {
                    reject(false);
                });
            });
        });
    }
}
var YTAudioFileFormat = (function () {
    function YTAudioFileFormat(data) {
        this.itag = Number(data.itag);
        this.container = data.container;
        this.audioEncoding = data.audioEncoding;
        this.audioBitrate = data.audioBitrate;
        this.url = data.url;
    }
    return YTAudioFileFormat;
}());
var YTAudioFileMeta = (function () {
    function YTAudioFileMeta(data) {
        this.title = data.title;
        this.author = data.author;
        this.length_seconds = Number(data.length_seconds);
        this.description = data.description;
        this.view_count = Number(data.view_count);
        this.formats = {};
        for (var _i = 0, _a = data.formats; _i < _a.length; _i++) {
            var format = _a[_i];
            this.formats[format.itag] = format;
        }
    }
    return YTAudioFileMeta;
}());


index.ts
#!/usr/bin/env node

import ytdl = require('ytdl-core');
import program = require('commander');
import fs = require('fs');
import ProgressBar = require('progress');
import Promise = require('bluebird');
require('console.table');

/*
 * Takumi Adachi
 * 02/01/2017
 *
 * MIT License
 * 
 * Node application to quickly and conveniently download adequate quality music, podcasts or audio from Youtube.
 * 
 * Has options for batch and checking bitrate of youtube audio.
 */

// Run main routine,
main();

// Main routine.
function main() {
    program
        .version('0.2.4')
        .description(
        `audl - A convenient node command-line app to download youtube audio content such as podcasts and music.

            Examples:

            audl -d https://www.youtube.com/watch?v=9bZkp7q19f0
            audl -i https://www.youtube.com/watch?v=9bZkp7q19f0
            audl -b batch.txt
            audl -I batch.txt`
        )
        .option('-d, --url [url]', 'Specify youtube link to download')
        .option('-b, --batch [file]', 'Specify a batch text file of youtube urls (LR separated) and download them all')
        .option('-i, --info [url]', 'Get list of quality options for that youtube content')
        .option('-I, --batch_info [file]', 'Specify a batch text file and get audio quality info of all youtube urls')
        /* Implement option -q --quality later */
        // .option('-q, --quality [id]',
        // `Specify quality and download youtube audio content. (Default is 140).

        // More info: https://en.wikipedia.org/wiki/YouTube#Quality_and_formats
        // [id]
        // 140	 M4A  AAC  128'bps
        // 141	 M4A  AAC  256'bps (No longer available.)
        // `, '140')
        .parse(process.argv);

    // Check if url is a valid youtube link.
    function valid_youtube_match(url) {
        let youtube_valid_regexp = /^.*(youtu.be\/|v\/|u\/\w\/|embed\/|watch\?v=|\&v=|\?v=)([^#\&\?]*).*/;
        let match = url.match(youtube_valid_regexp);
        if (match && match[2].length == 11) {
            return true;
        } else {
            return false;
        }

    }

    // Download youtube content as audio.
    if (program.url) {
        let url = program.url;
        if (!valid_youtube_match(url)) {
            console.log(program.url + 'is not a valid youtube url.');
            return
        }
        
        console.log();
        YTdownloadAsAudio(url).then((result) => {
        }).error((err) => {
            console.log(err);
        })

        return;
    }

    // Get itag info and output to console.
    if (program.info) {
        let url = program.info;
        // Check if valid url.
        if (!valid_youtube_match(url)) {
            console.log(url + 'is not a valid youtube url.');
            return
        }

        getInfo(url).then((info) => {
            let audio_file_meta = new YTAudioFileMeta(info);
            let title = info.title.replace(/[^a-z]+/gi, '_').toLowerCase(); // music_title.
            let formats = audio_file_meta.formats;
            let itag_info = [];

            function push(itag) {
                itag_info.push({
                    title: title,
                    itag: formats[itag].itag.toString(),
                    encoding: formats[itag].audioEncoding.toString(),
                    bitrate: formats[itag].audioBitrate.toString(),
                })
            }

            for (let itag in formats) {
                let found = false;
                if (itag === '139') push(itag)
                if (itag === '140') push(itag)
                if (itag === '141') push(itag)
            }

            console.log()
            console.table(itag_info);
        }).error((err) => {
            console.log(err);
        });

        return;
    }

    // Get Itag info for a batch of youtube urls.
    if (program.batch_info) {
        // Example: Read from a file of youtube links separated by linefeeds.
        var youtube_urls = fs.readFileSync(program.batch_info).toString().split("\r");
        let getInfo_promises = [];
        let itag_info = [];

        for (let url of youtube_urls) {
            // Remove \r \n from string before validating.
            url = url.replace(/(\r\n|\n|\r)/gm, "");
            if (!valid_youtube_match(url)) {
                console.log(url + ' is not a valid youtube url.');
                continue
            }

            let promise = getInfo(url).then((info) => {
                let audio_file_meta = new YTAudioFileMeta(info);
                let title = info.title.replace(/[^a-z]+/gi, '_').toLowerCase(); // music_title.
                let formats = audio_file_meta.formats;

                function push(itag) {
                    itag_info.push({
                        title: title,
                        itag: formats[itag].itag.toString(),
                        encoding: formats[itag].audioEncoding.toString(),
                        bitrate: formats[itag].audioBitrate.toString(),
                        url: url
                    });
                }

                for (let itag in formats) {
                    if (itag === '139') push(itag);
                    if (itag === '140') push(itag);
                    if (itag === '141') push(itag);
                }

                return itag_info;
            })

            getInfo_promises.push(promise);
        }

        Promise.map(getInfo_promises, (info) => {
            // Do only one request at a time using concurrency.
        }, { concurrency: 1 }).then(function () {
            console.log();
            console.table(itag_info);
        }).error((err) => {
            console.log(err);
        });;

        return;
    }

    // Download a batch of youtube urls in audio format.
    if (program.batch) {
        // Example: Read from a file of youtube links separated by linefeeds.
        var youtube_urls = fs.readFileSync(program.batch).toString().split("\n");
        let download_promises = [];
        let itag_info = [];

        for (let url of youtube_urls) {
            // Remove \r \n from string before validating.
            url = url.replace(/(\r\n|\n|\r)/gm, "");
            // Check if valid youtube url.
            if (!valid_youtube_match(url)) {
                console.log(url + 'is not a valid youtube url and will not be downloaded.');
                continue
            }
            let promise = YTdownloadAsAudio(url);

            download_promises.push(promise);
        }

        Promise.map(download_promises, (result) => {
            // Do only one request at a time using concurrency.
        }, { concurrency: 1 }).then(function (result) {

        }).error((err) => {
            console.log(err);
        });

        return;
    }

    // If program was called with no arguments, show help.
    if (program.args.length === 0) program.help();

    // Return a promise to get youtube info such as itag, quality, bitrate.
    function getInfo(url) {
        return new Promise(function (resolve, reject) {
            ytdl.getInfo(url, function (err, info) {
                if (info) resolve(info)
                if (err) reject(err);
            })
        })
    }

    // Return a promise to download youtube audio content.
    function YTdownloadAsAudio(url) {
        return new Promise(function (resolve, reject) {
            ytdl.getInfo(url, function (err, info) {
                let audio_file_meta = new YTAudioFileMeta(info);
                let file_type = '.m4a';
                // Remove unneeded characters and replace with underscores for readability and make it file friendly.
                let file_name = audio_file_meta.title.replace(/[^a-z]+/gi, '_').toLowerCase() + file_type; // music_title.m4a
                let write_stream = fs.createWriteStream(file_name);
                let audio = ytdl(url, { quality: 140 })
                audio.pipe(write_stream);
                audio.on('response', function (res) {
                    let dataRead = 0;
                    let totalSize = parseInt(res.headers['content-length']);
                    let options = {
                        complete: '\u001b[42m \u001b[0m', // Green.
                        incomplete: '\u001b[41m \u001b[0m', // Red.
                        width: 20,
                        total: totalSize
                    }
                    //Example download youtube audio and save as .m4a with a progressbar.
                    let bar = new ProgressBar(` downloading [:bar] :percent :etas :current/:total - ${file_name}`, options);

                    res.on('data', function (data) {
                        let chunk = data.length;
                        bar.tick(chunk)
                    })
                });
                audio.on('finish', () => {
                    resolve(true);
                });
                audio.on('error', () => {
                    reject(false);
                });

            })
        });

    }


}

interface AudioFileFormatsInterface {
    itag: number;
    container: string;
    audioEncoding: string;
    audioBitrate: string;
    url: string;
}

interface AudioFileMetaInterface {
    title: string;
    author: string;
    length_seconds: number;
    description: string;
    view_count: number;
    formats: any;
}

class YTAudioFileFormat implements AudioFileFormatsInterface {
    itag: number;
    container: string;
    audioEncoding: string;
    audioBitrate: string;
    url: string;

    constructor(data: any) {
        this.itag = Number(data.itag);
        this.container = data.container;
        this.audioEncoding = data.audioEncoding;
        this.audioBitrate = data.audioBitrate;
        this.url = data.url;
    }
}

class YTAudioFileMeta implements AudioFileMetaInterface {
    title: string;
    author: string;
    length_seconds: number;
    description: string;
    view_count: number;
    formats: any;

    constructor(data: any) {
        this.title = data.title;
        this.author = data.author;
        this.length_seconds = Number(data.length_seconds);
        this.description = data.description;
        this.view_count = Number(data.view_count);
        this.formats = {};
        for (let format of data.formats) {
            this.formats[format.itag] = format;
        }
    }
}